[#5] 로그인 기능 추가

cold-pumpkin opened the pull request:
Closes #5 
Closes #7 

### 로그인 관련 기능 구현 및 예외 처리
- [X] 로그인 시 이메일로 회원 정보 조회하기
- [X] 회원 정보 없을 시 예외 처리
- [X] 회원 정보 존재할 시 패스워드 비교
- [X] 패스워드 틀린 경우 예외 처리
- [X] 로그인 성공할 시 최근 로그인 시간 업데이트
- [X] 로그인 관련 테스트 코드 작성

cold-pumpkin committed:
[#5] 로그인 관련 매퍼 작성

- 비밀번호 체크 전에 이메일로 유저 정보가 존재하는지 찾기 위한 쿼리 작성
- 로그인 성공 시 마지막 로그인 시간 업데이트 쿼리 작성
- Java LocalDateTime 타입과 매핑을 위해 typeHandler 라이브러리 및
  MyBatis 설정파일에 설정 추가

cold-pumpkin committed:
[#5] 모델에 도메일 관련 로직 추가

- 로그인 시 마지막 로그인 시간 업데이트를 위한 메서드 추가
- MyBatis에서 유저 정보 조회 시 매핑을 위한 생성자 명시
- Email 클래스 address 필드 불변성 보장을 위해 final 추가

cold-pumpkin committed:
[#5] 로그인 관련 서비스 및 커스텀 예외 추가

- 메일로 유저 정보 찾기 및 로그인 메서드 추가
- 유저 정보 찾지 못했을 경우 NotFoundException 예외 처리
- 패스워드 틀렸을 경우 InvalidPasswordException 예외 처리

cold-pumpkin committed:
[#5] 로그인 요청 처리 컨트롤러 추가

cold-pumpkin committed:
[#5] 로그인 관련 테스트 로직 추가

f-lab-dev commented on pom.xml:
```diff
@@ -34,6 +35,12 @@
 
+        <dependency>
+            <groupId>org.mybatis</groupId>
+            <artifactId>mybatis-typehandlers-jsr310</artifactId>
```
이것을 추가한 의도도 주석에 써있으면 좋을 것 같습니다~

cold-pumpkin replied:
네 주석으로 추가하겠습니다~

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/controller/user/UserRestController.java:
```diff
@@ -21,4 +23,10 @@ public JoinResponse join(@RequestBody JoinRequest joinRequest) {
+    @GetMapping(path = "/login")
+    public LoginResponse login(@RequestBody LoginRequest loginRequest) {
+        User user = userService.login(new Email(loginRequest.getEmail()), loginRequest.getPassword());
+        return new LoginResponse(user);
```
`new` 키워드를 쓰기보다는 static으로 `from` 메소드를 써주는게 좋습니다. 이런 정적 팩토리 메소드를 왜 써야하는지 한번 알아보시면 좋겠습니다~

cold-pumpkin replied:
생성자 대신 정적 팩토리 메소드를 활용할 때 얻을 수 있는 장점은 아래와 같습니다. 
(Effective Java를 참고했습니다!)
1. 메소드의 네이밍을 통해 객체 생성시 의미를 드러낼 수 있습니다.
   - 생성자에 매개변수를 넘기는 것만 보고서는 반환될 객체의 특성을 제대로 파악할 수 없지만, 정적 팩토리 메소드를 사용하면 생성될 객체에 대해 설명할 수 있는 메소드의 이름을 붙여 사용할 수 있습니다.
2. 호출할 때마다 인스턴스를 새로 생성하지 않아도 됩니다.
   - 불변 클래스의 경우 객체를 매번 생성하는 것이 아니라, 한번만 생성해 재활용함으로써 객체 생성에 대한 비용을 줄일 수 있습니다.  
3. 반환 타입의 하위 타입 객체를 반환할 수 있습니다.
   - 내부적으로 객체를 반환할 때 해당 클래스의 하위 타입을 자유롭게 선택함으로써 설계의 유연성을 높일 수 있습니다. 이 때 실제 구현체들은  내부에 숨긴 상태로 객체를 반환할 수 있기 때문에 API의 외견을 작게 유지할 수 있습니다.
4. 매개변수에 따라 다른 클래스의 객체를 반환할 수 있습니다.
   - 스태틱 메소드에 넘겨주는 변수에 따라 해당 변수에 매핑된 하위 타입 객체를 반환할 수 있습니다. 이 역시 OCP에 기반한 유연한 확장을 가능하게 해줍니다.
5. 정적 팩토리 메소드를 작성할 시점에 반환할 객체의 클래스가 존재하지 않아도 됩니다.
   - 예를 들어 반환 타입이 인터페이스인 경우, 클라이언트가 명시한 조건에 따라 그에 해당하는 구현체가 런타임 시점에 동적으로 생성되도록 할 수 있습니다. 이 역시 유연한 설계를 가능하게 해줍니다.

추후 우리 프로젝트에서도 비즈니스 로직이 더 복잡해지고 확장됨에 따라서 이런 정적 팩토리의 장점들을 활용해 볼 수 있을 것 같습니다. 일반적으로 매개 변수가 하나인 경우 `from`을 사용하기 때문에 `from` 정적 팩토리 메소드를 추가해보겠습니다~

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/model/user/User.java:
```diff
@@ -43,4 +42,35 @@
 
     private LocalDateTime updateAt;
 
+    public void checkPassword(String inputPw, String userPw) {
```
1. `checkPassword`라는 이름에서는 **패스워드가 동일한지 확인한다** 라는 행위가 명확하게 보이지 않습니다~
2. 서버에서 굳이 패스워드가 맞는지 검사해줄 필요가 있을까요? 클라이언트 화면에 **비밀번호 확인** 이라는 필드가 애초에 왜 존재하는지 의도를 고민해보시면 좋을 것 같네요~

cold-pumpkin replied:
생각해보니 굳이 DB에서 유저 정보를 모두 가져와 사용자가 입력한 패스워드를 비교하지 않아도, 입력한 이메일과 패스워드에 해당하는 데이터가 존재하는지 확인하는 것 만으로도 로그인 처리를 할 수 있을 것 같네요!
이 방식으로 변경해 보겠습니다~

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/UserService.java:
```diff
@@ -43,4 +46,20 @@ public User join(Email email, String nickname, String password) {
     }
 
+    @Transactional
+    public User login(Email email, String password) {
```
이런 인증/인가와 관련된 코드는 `UserService`의 책임이 맞을까요? `UserService`는 어떤 책임을 갖고있는지, 그 책임에 맞는 메소드인지 고민해보시면 좋겠습니다~

cold-pumpkin replied:
단일책임원칙을 지키기 위해 `UserService`는 유저 정보 조회, 정보 삭제(탈퇴), 정보 변경 등 유저와 직접적인 관련이 있는 작업들만 두는 것이 좋을 것 같습니다!
로그인, 로그아웃 같은 작업들은 `LoginService`로 따로 분리하는 방식으로 수정해보겠습니다~

f-lab-dev commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -4,9 +4,44 @@
         VALUES(#{id}, #{email.address}, #{nickname}, #{password}, #{status}, #{lastLoginAt}, #{createAt}, #{updateAt})
     </insert>
 
+    <update id="updateUser" parameterType="User">
```
이 쿼리는 로그인이라는 기능에 필요한 작업내용이였을까요?

cold-pumpkin replied:
엄 요것과 관련해 생각해보다보니 궁금한게 생겼습니다..!
유저가 로그인한 직후에 항상 마지막 로그인 시간(`last_login_at` 칼럼)을 업데이트하기 때문에 로그인 프로세스의 하나로 생각했었는데요, 그렇다 하더라도 이런 경우 로그인 기능과는 별도로 보는 것이 맞을까요~?
그리고 이런 범용적인 업데이트 쿼리를 만들어 사용하면 나중에 로그인 직후 변경할 필드가 생성/추가되거나(ex. 마지막 로그인 시간 갱신과 함께 유저 로그인 횟수도 카운트) 로그인 외에 유저 정보 변경 기능 등에서도 함께 활용할 수 있어 편할거 같다고 생각했는데, 그렇다 해도 명확한 의도가 나타나도록 마지막 로그인 시간 칼럼만 업데이트하는 쿼리를 따로 두는 것이 좋을까요?

f-lab-dev commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -4,9 +4,44 @@
+    </update>
+
+    <select id="findByEmail" parameterType="Email" resultMap="userMap">
+        SELECT * FROM user WHERE email = #{address}
```
`*`을 쓰기보다는 유지보수성을 위해 어떤 컬럼들을 필요로하는지 명확하게 써주시는게 좋습니다~

cold-pumpkin replied:
넵~! 👌

cold-pumpkin committed:
[#5] 리뷰반영: 라이브러리 추가 관련 주석

- MyBatis TypHandler가 Java 8에 추가된 java.time 패키지의 API 인식하려면
  mybatis-typehandlers-jsr310 라이브러리 필요
- MyBatis 3.4 버전 이상부터 활용 가능

cold-pumpkin committed:
[#5] 리뷰반영: 로그인 관련 서비스 분리

- 단일책임원칙을 고려해 로그인 관련 서비스는 LoginService로 분리
- 로그인 시 유저 이메일과 패스워드로 DB에서 데이터를 찾아 리턴하는
  방식으로 변경

cold-pumpkin committed:
[#5] 리뷰반영: 생성자 대신 정적 팩토리 메소드 활용

cold-pumpkin committed:
[#5] 쿼리 가독성 향상을 위한 들여쓰기 수정

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/LoginService.java:
```diff
@@ -0,0 +1,30 @@
+
+@Service
+@RequiredArgsConstructor
+public class LoginService {
```
로그인도 세션 방식 로그인, 토큰 방식 로그인 등 여러가지 방식으로 사용될 수 있을 것 같습니다~
추상화를 하여 확장할 수 있도록 해주면 어떨까요?

cold-pumpkin replied:
넵 그게 좋을 것 같습니다! LoginService는 인터페이스로 두고, 다음 세션 스토리지(Redis)를 활용한 로그인 방식을 따로 구현해보겠습니다~

cold-pumpkin committed:
[#5] 리뷰반영: 로그인 관련 서비스 추상화

- 로그인 방식 변경 및 확장을 고려해 인터페이스 레이어 추가
- BasicLoginService(세션없는 기본 로그인 방식) 구현
- 추후 Redis 활용한 로그인 서비스 구현 예정

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/BasicLoginService.java:
```diff
@@ -0,0 +1,30 @@
+
+@Service
+@RequiredArgsConstructor
+public class BasicLoginService implements LoginService {
```
1. `BasicLoginService`은 이 클래스 내부에서 어떤 방식으로 로그인을 하는지 명확하게 보이고있나요?
2. 이 클래스는 세션방식으로 로그인을 하는 책임을 갖고있는 것 같은데 이 클래스 내부에서 세션의 존재는 보이지 않습니다. 그러면 세션을 컨트롤하는 책임을 다른 클래스가 갖고있다는 것 아닐까요?

cold-pumpkin replied:
기존에 로그인 서비스를 구현할 때 세션 방식을 미처 추가하지 못했었는데요, 이 기존 코드를 `BasicLoginService` 라는 이름으로 남겨두고 세션을 활용한 방식을 따로 구현해 나중에 올리려고 했는데 불명확한 네이밍 때문에 오히려 혼란을 빚은 것 같습니다 ㅠ
좀 더 명확히 하기 위해 이 `BasicLoginService`는 삭제하고, 실제 서비스에서 사용할 세션 기반의 로그인 서비스를 구현해 다시 올리도록 하겠습니다~

cold-pumpkin committed:
[#7] 세션 기반 로그인 구현을 위해 Redis 추가

cold-pumpkin committed:
[#7] 세션 활용하지 않은 기존의 로그인 방식 제거

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/SessionLoginService.java:
```diff
@@ -0,0 +1,45 @@
+import static org.apache.logging.log4j.util.Strings.isNotEmpty;
+
+@Slf4j
+@Primary
```
이거는 굳이 붙여주신 이유가 있으실까요?

cold-pumpkin replied:
아이고 기존에 `LoginService` 인터페이스를 구현한 클래스가 `BasicLoginService`와 `SessionLoginService` 두 개였는데요, 그 중 `SessionLoginService`를 사용하기 위해 `@Primary`를  붙였었습니다..! 이젠 SessionLoginService 하나만 있으니까 삭제해도 되는데 미처 못지웠네요! 
Slf4j도 개발 중에 사용했던거라 삭제하겠습니다~

cold-pumpkin committed:
[#7] 리뷰반영: 불필요한 애노테이션 삭제
